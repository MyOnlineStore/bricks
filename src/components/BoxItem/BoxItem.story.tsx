import { boolean, select, text } from '@storybook/addon-knobs/react';
import { storiesOf } from '@storybook/react';
import React from 'react';
import styled from 'styled-components';
import Box from '../Box';
import Spacer from '../Spacer/Spacer.style';
import Text from '../Text';
import BoxItem from './BoxItem.style';

const Wrapper = styled.div`
    border: 1px dashed #aaa;
    height: 60vh;
    padding: 6px;
`;

const StyledBoxItem = styled(BoxItem)`
    background: #eee;
    border: 1px dashed #aaa;
    padding: 6px 0;
`;

const StyledContent = styled(Text)`
    border: 1px dashed #aaa;
    margin: 6px 12px;
    padding: 12px;
`;

storiesOf('BoxItem', module)
    .add('Default', () => {
        const verticalOptions = ['top', 'bottom', 'center', 'stretch'];
        const horizontalOptions = ['left', 'right', 'center', 'stretch'];
        const gutterOptions = [[0], [12], [24]];
        const items:Array<JSX.Element> = [];
        const boxgroupId = 'BOX';
        const boxItemgroupId = 'BOXITEM';

        for (let i = 0; i < 3; i++) {
            items.push(
                <StyledBoxItem
                    key={i}
                    grow={boolean('grow', true)}
                    shrink={boolean('shrink', true)}
                    size={text('size', 'auto')}
                    gutter={[6]}
                    horizontal={select('horizontal', horizontalOptions, horizontalOptions[0], boxItemgroupId)}
                    vertical={select('vertical', verticalOptions, verticalOptions[0], boxItemgroupId)}
                >
                    <StyledContent>BoxItem content</StyledContent>
                    <StyledContent>BoxItem content</StyledContent>
                    <StyledContent>BoxItem content</StyledContent>
                </StyledBoxItem>
            );
        }

        return (
            <Wrapper>
                <Box
                    horizontal={select('Box horizontal', horizontalOptions, horizontalOptions[0], boxgroupId)}
                    vertical={select('Box vertical', verticalOptions, verticalOptions[0], boxgroupId)}
                    wrap={true}
                >
                    {items}
                </Box>
            </Wrapper>
        );
    });
