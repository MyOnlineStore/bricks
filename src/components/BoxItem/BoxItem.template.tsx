import React, { StatelessComponent } from 'react';
import { OffsetType } from '../../types/OffsetType';
import { StyledType } from '../../utility/styled';

type PropsType = StyledType & {
    size:string;
    grow?:boolean;
    shrink?:boolean;
    gutter?:Array<OffsetType>;
    vertical?:'top'|'bottom'|'center'|'stretch';
    horizontal?:'left'|'right'|'center'|'stretch';
};

const BoxItem:StatelessComponent<PropsType> = (props):JSX.Element => {
    return <div className={props.className}>{props.children}</div>;
};

export { PropsType };
export default BoxItem;
