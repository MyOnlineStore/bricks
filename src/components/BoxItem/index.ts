export { default as default } from './BoxItem.style';
