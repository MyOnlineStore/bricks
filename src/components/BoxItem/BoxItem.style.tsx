import { StyledComponentClass as _S } from 'styled-components';
import styled from '../../utility/styled';
import BoxItem from './BoxItem.template';
import formatOffset from './formatters/formatOffset';

const StyledBoxItem = styled(BoxItem)`
    display: flex;
    flex-flow: column nowrap;
    justify-content: ${({ vertical }):string => {
        switch (vertical) {
            case undefined:
            case 'top': return 'flex-start';
            case 'bottom': return 'flex-end';
            default: return vertical;
        }
    }};
    align-items: ${({ horizontal }):string => {
        switch (horizontal) {
            case undefined:
            case 'left': return 'flex-start';
            case 'right': return 'flex-end';
            default: return horizontal;
        }
    }};
    align-content: ${({ horizontal }):string => {
        switch (horizontal) {
            case undefined:
            case 'left': return 'flex-start';
            case 'right': return 'flex-end';
            default: return horizontal;
        }
    }};

    && {
        flex-grow: ${({ grow }):string => grow ? '1' : '0'};
        flex-shrink: ${({ shrink }):string => shrink ? '1' : '0'};
        flex-basis: ${({ size }):string => size};
        margin: ${({ gutter }):string => gutter !== undefined ? formatOffset(gutter) : '0' };
    }

    > * {
        flex-grow: ${({ vertical }):number => vertical === 'stretch' ? 1 : 0};
        flex-shrink: 1;
    }
`;

export default StyledBoxItem;
