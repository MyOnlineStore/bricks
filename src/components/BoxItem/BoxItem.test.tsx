import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import React from 'react';
import BoxItem from '.';

describe('BoxItem', () => {
    it('aligns children to the right', () => {
        const component = shallow(<BoxItem size="auto" horizontal="right" />);

        expect(toJson(component)).toMatchSnapshot();
    });

    it('aligns children in the center', () => {
        const component = shallow(<BoxItem size="auto" horizontal="center" />);

        expect(toJson(component)).toMatchSnapshot();
    });

    it('aligns children to the bottom', () => {
        const component = shallow(<BoxItem size="auto" vertical="bottom" />);

        expect(toJson(component)).toMatchSnapshot();
    });

    it('aligns children in the vertical center', () => {
        const component = shallow(<BoxItem size="auto" vertical="center" />);

        expect(toJson(component)).toMatchSnapshot();
    });

    it('creates a gutter around it', () => {
        const component = shallow(<BoxItem size="auto" gutter={[0, 12, 24, 'auto']} />);

        expect(toJson(component)).toMatchSnapshot();
    });

    it('prevents wrapping of children', () => {
        const component = shallow(<BoxItem size="auto" />);

        expect(toJson(component)).toMatchSnapshot();
    });

    it('stretches the children', () => {
        const component = shallow(<BoxItem size="auto" horizontal="stretch" vertical="stretch" />);

        expect(toJson(component)).toMatchSnapshot();
    });
});
