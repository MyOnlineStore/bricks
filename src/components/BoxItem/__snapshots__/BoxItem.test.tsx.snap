// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BoxItem aligns children in the center 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-flow: column nowrap;
  -ms-flex-flow: column nowrap;
  flex-flow: column nowrap;
  -webkit-box-pack: start;
  -webkit-justify-content: flex-start;
  -ms-flex-pack: start;
  justify-content: flex-start;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-align-content: center;
  -ms-flex-line-pack: center;
  align-content: center;
}

.c0.c0 {
  -webkit-box-flex: 0;
  -webkit-flex-grow: 0;
  -ms-flex-positive: 0;
  flex-grow: 0;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
  -webkit-flex-basis: auto;
  -ms-flex-preferred-size: auto;
  flex-basis: auto;
  margin: 0;
}

.c0 > * {
  -webkit-box-flex: 0;
  -webkit-flex-grow: 0;
  -ms-flex-positive: 0;
  flex-grow: 0;
  -webkit-flex-shrink: 1;
  -ms-flex-negative: 1;
  flex-shrink: 1;
}

<BoxItem
  className="c0"
  horizontal="center"
  size="auto"
/>
`;

exports[`BoxItem aligns children in the vertical center 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-flow: column nowrap;
  -ms-flex-flow: column nowrap;
  flex-flow: column nowrap;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: flex-start;
  -webkit-box-align: flex-start;
  -ms-flex-align: flex-start;
  align-items: flex-start;
  -webkit-align-content: flex-start;
  -ms-flex-line-pack: start;
  align-content: flex-start;
}

.c0.c0 {
  -webkit-box-flex: 0;
  -webkit-flex-grow: 0;
  -ms-flex-positive: 0;
  flex-grow: 0;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
  -webkit-flex-basis: auto;
  -ms-flex-preferred-size: auto;
  flex-basis: auto;
  margin: 0;
}

.c0 > * {
  -webkit-box-flex: 0;
  -webkit-flex-grow: 0;
  -ms-flex-positive: 0;
  flex-grow: 0;
  -webkit-flex-shrink: 1;
  -ms-flex-negative: 1;
  flex-shrink: 1;
}

<BoxItem
  className="c0"
  size="auto"
  vertical="center"
/>
`;

exports[`BoxItem aligns children to the bottom 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-flow: column nowrap;
  -ms-flex-flow: column nowrap;
  flex-flow: column nowrap;
  -webkit-box-pack: end;
  -webkit-justify-content: flex-end;
  -ms-flex-pack: end;
  justify-content: flex-end;
  -webkit-align-items: flex-start;
  -webkit-box-align: flex-start;
  -ms-flex-align: flex-start;
  align-items: flex-start;
  -webkit-align-content: flex-start;
  -ms-flex-line-pack: start;
  align-content: flex-start;
}

.c0.c0 {
  -webkit-box-flex: 0;
  -webkit-flex-grow: 0;
  -ms-flex-positive: 0;
  flex-grow: 0;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
  -webkit-flex-basis: auto;
  -ms-flex-preferred-size: auto;
  flex-basis: auto;
  margin: 0;
}

.c0 > * {
  -webkit-box-flex: 0;
  -webkit-flex-grow: 0;
  -ms-flex-positive: 0;
  flex-grow: 0;
  -webkit-flex-shrink: 1;
  -ms-flex-negative: 1;
  flex-shrink: 1;
}

<BoxItem
  className="c0"
  size="auto"
  vertical="bottom"
/>
`;

exports[`BoxItem aligns children to the right 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-flow: column nowrap;
  -ms-flex-flow: column nowrap;
  flex-flow: column nowrap;
  -webkit-box-pack: start;
  -webkit-justify-content: flex-start;
  -ms-flex-pack: start;
  justify-content: flex-start;
  -webkit-align-items: flex-end;
  -webkit-box-align: flex-end;
  -ms-flex-align: flex-end;
  align-items: flex-end;
  -webkit-align-content: flex-end;
  -ms-flex-line-pack: end;
  align-content: flex-end;
}

.c0.c0 {
  -webkit-box-flex: 0;
  -webkit-flex-grow: 0;
  -ms-flex-positive: 0;
  flex-grow: 0;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
  -webkit-flex-basis: auto;
  -ms-flex-preferred-size: auto;
  flex-basis: auto;
  margin: 0;
}

.c0 > * {
  -webkit-box-flex: 0;
  -webkit-flex-grow: 0;
  -ms-flex-positive: 0;
  flex-grow: 0;
  -webkit-flex-shrink: 1;
  -ms-flex-negative: 1;
  flex-shrink: 1;
}

<BoxItem
  className="c0"
  horizontal="right"
  size="auto"
/>
`;

exports[`BoxItem creates a gutter around it 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-flow: column nowrap;
  -ms-flex-flow: column nowrap;
  flex-flow: column nowrap;
  -webkit-box-pack: start;
  -webkit-justify-content: flex-start;
  -ms-flex-pack: start;
  justify-content: flex-start;
  -webkit-align-items: flex-start;
  -webkit-box-align: flex-start;
  -ms-flex-align: flex-start;
  align-items: flex-start;
  -webkit-align-content: flex-start;
  -ms-flex-line-pack: start;
  align-content: flex-start;
}

.c0.c0 {
  -webkit-box-flex: 0;
  -webkit-flex-grow: 0;
  -ms-flex-positive: 0;
  flex-grow: 0;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
  -webkit-flex-basis: auto;
  -ms-flex-preferred-size: auto;
  flex-basis: auto;
  margin: 0px 12px 24px auto;
}

.c0 > * {
  -webkit-box-flex: 0;
  -webkit-flex-grow: 0;
  -ms-flex-positive: 0;
  flex-grow: 0;
  -webkit-flex-shrink: 1;
  -ms-flex-negative: 1;
  flex-shrink: 1;
}

<BoxItem
  className="c0"
  gutter={
    Array [
      0,
      12,
      24,
      "auto",
    ]
  }
  size="auto"
/>
`;

exports[`BoxItem prevents wrapping of children 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-flow: column nowrap;
  -ms-flex-flow: column nowrap;
  flex-flow: column nowrap;
  -webkit-box-pack: start;
  -webkit-justify-content: flex-start;
  -ms-flex-pack: start;
  justify-content: flex-start;
  -webkit-align-items: flex-start;
  -webkit-box-align: flex-start;
  -ms-flex-align: flex-start;
  align-items: flex-start;
  -webkit-align-content: flex-start;
  -ms-flex-line-pack: start;
  align-content: flex-start;
}

.c0.c0 {
  -webkit-box-flex: 0;
  -webkit-flex-grow: 0;
  -ms-flex-positive: 0;
  flex-grow: 0;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
  -webkit-flex-basis: auto;
  -ms-flex-preferred-size: auto;
  flex-basis: auto;
  margin: 0;
}

.c0 > * {
  -webkit-box-flex: 0;
  -webkit-flex-grow: 0;
  -ms-flex-positive: 0;
  flex-grow: 0;
  -webkit-flex-shrink: 1;
  -ms-flex-negative: 1;
  flex-shrink: 1;
}

<BoxItem
  className="c0"
  size="auto"
/>
`;

exports[`BoxItem stretches the children 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-flow: column nowrap;
  -ms-flex-flow: column nowrap;
  flex-flow: column nowrap;
  -webkit-box-pack: stretch;
  -webkit-justify-content: stretch;
  -ms-flex-pack: stretch;
  justify-content: stretch;
  -webkit-align-items: stretch;
  -webkit-box-align: stretch;
  -ms-flex-align: stretch;
  align-items: stretch;
  -webkit-align-content: stretch;
  -ms-flex-line-pack: stretch;
  align-content: stretch;
}

.c0.c0 {
  -webkit-box-flex: 0;
  -webkit-flex-grow: 0;
  -ms-flex-positive: 0;
  flex-grow: 0;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
  -webkit-flex-basis: auto;
  -ms-flex-preferred-size: auto;
  flex-basis: auto;
  margin: 0;
}

.c0 > * {
  -webkit-box-flex: 1;
  -webkit-flex-grow: 1;
  -ms-flex-positive: 1;
  flex-grow: 1;
  -webkit-flex-shrink: 1;
  -ms-flex-negative: 1;
  flex-shrink: 1;
}

<BoxItem
  className="c0"
  horizontal="stretch"
  size="auto"
  vertical="stretch"
/>
`;
